### ISSUE TITLE
Using nested plist::Dictionary with serde?

### ISSUE THREAD

**Issue by madig** [ebarnard/rust-plist#54]
The following code does not compile. Is there a way to use `plist::Dictionary` in a nested way?

```rust
use plist;
#[macro_use]
extern crate serde_derive;

fn main() {
    #[derive(Deserialize)]
    struct LayerinfoData {
        color: Option<String>,
        lib: Option<plist::Dictionary>,
    }

    let lib_dict: LayerinfoData = plist::from_bytes(r#"
    <?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>color</key>
    <string>1,0.75,0,0.7</string>
    <key>lib</key>
    <dict>
      <key>com.typemytype.robofont.segmentType</key>
      <string>curve</string>
    </dict>
  </dict>
</plist>
"#.as_bytes()).unwrap();

    println!("{:?}, {:?}.", lib_dict.color, lib_dict.lib);
}
```

Compiler output:
```
error[E0277]: the trait bound `plist::dictionary::Dictionary: fontinfo::_IMPL_DESERIALIZE_FOR_FontInfo::_serde::Deserialize<'_>` is not satisfied
  --> src/layer.rs:75:17
   |
75 |                 lib: Option<plist::Dictionary>,
   |                 ^^^ the trait `fontinfo::_IMPL_DESERIALIZE_FOR_FontInfo::_serde::Deserialize<'_>` is not implemented for `plist::dictionary::Dictionary`
   |
   = note: required because of the requirements on the impl of `fontinfo::_IMPL_DESERIALIZE_FOR_FontInfo::_serde::Deserialize<'_>` for `std::option::Option<plist::dictionary::Dictionary>`
   = note: required by `fontinfo::_IMPL_DESERIALIZE_FOR_FontInfo::_serde::de::SeqAccess::next_element`

error[E0277]: the trait bound `plist::dictionary::Dictionary: fontinfo::_IMPL_DESERIALIZE_FOR_FontInfo::_serde::Deserialize<'_>` is not satisfied
  --> src/layer.rs:75:17
   |
75 |                 lib: Option<plist::Dictionary>,
   |                 ^^^ the trait `fontinfo::_IMPL_DESERIALIZE_FOR_FontInfo::_serde::Deserialize<'_>` is not implemented for `plist::dictionary::Dictionary`
   |
   = note: required because of the requirements on the impl of `fontinfo::_IMPL_DESERIALIZE_FOR_FontInfo::_serde::Deserialize<'_>` for `std::option::Option<plist::dictionary::Dictionary>`
   = note: required by `fontinfo::_IMPL_DESERIALIZE_FOR_FontInfo::_serde::de::MapAccess::next_value`
```

**Comment by ebarnard** [comment]
Unfortunately `plist::Dictionary` doesn't, at the moment, implement `serde::Deserialize`.

If you really have a plist file containing a heterogeneous collection of objects you can deserialize it into a `HashMap<String, serde_json::Value>`, although it is a bit of a hack.

**Comment by cmyr** [comment]
I'm bumping into something similar, and looking through the code i see that there was at least an attempt to implement serialize and deserialize for `Dictionary`. I'm curious if the decision to not follow through with this was because of some technical issue, or something else? If you would be open to having it implemented I'd be happy to dig into it...

**Comment by ebarnard** [comment]
Most of the Dictionary code is copy-pasted from serde_json. Implementing (de)serialize for Dictionary would mean implementing it for Value, which I’ve never got around to doing.

**Comment by cmyr** [comment]
okay, good to know. I'll look into getting that implemented next time I have the proper combination of energy and exasperation. 😁

**Comment by madig** [comment]
I think I'm running into this again. I'm trying to convert a `serde_json::Value` to a `plist::Value`. is there any workaround? A cross-deserialization like 

```rust
let data =
    serde_json::from_str::<plist::Value>(&serde_json::to_string(&value).unwrap())
        .unwrap();
```

Needs `Deserialize` for `plist::Value`.

**Comment by cmyr** [comment]
I think the best approach would be to write your own conversion method that converts the various `serde_json` types to the appropriate `plist` types? Like,

```rust
fn convert(from: serde_json::Value) -> plist::Value {
    match from {
        serde_json::Value::Bool(val) => plist::Value::Boolean(val),
        serde_json::Value::Number(num) => {
            if let Some(float) = num.as_f64() {
                plist::Real(float)
            } else if let Some(uint) = num.as_u64() {
                plist::Number(uint.into())
            } else if let Some(int) = num.as_u64() {
                plist::Number(int.into())
            } else {
                unreachable!()
            }
        }
        // etc
    }
}

**Comment by madig** [comment]
Hm, yes, an be recursive for containers I guess...

**Comment by ebarnard** [comment]
Not the most elegant solution, but you could try it the other way round.

```rust
let mut cursor = std::io::Cursor::new(Vec::new());
plist::to_writer_binary(&mut cursor, &value).unwrap()
cursor.set_position(0);
plist::Value::from_reader(&mut cursor);
```

**Comment by madig** [comment]
That worked, thanks! Still better than writing my own converter :)

**Comment by ctrlcctrlv** [comment]
Thank you very much for that recently pushed work @MLKrisJohnson, this was a blocker on proper data type of `glifparser::Glif::lib`, which is now a `plist::Dictionary` as it always should've been: https://github.com/MFEK/glifparser.rlib/commit/69389e67f536d7f9685313d89b7510cd88263a1b

I imagine that your work will broadly help anyone dealing with formats that have inline plist's like UFO .glif

**Comment by DiegoMagdaleno** [comment]
Hi, has this issue being fixed?.

If so, I suggest closing it, as it is confusing if it is supported.

Thank you :)

**Comment by ctrlcctrlv** [comment]
I suggest leaving it open until it's actually in a release. I'm tracking HEAD until then.

**Comment by ebarnard** [comment]
Released in v1.3.0.
